source/common/network/lc_trie.h

```diff
@@ -0,0 +1,487 @@
+#pragma once
+
+#include <arpa/inet.h>
+
+#include <algorithm>
+#include <vector>
+
+#include "envoy/common/exception.h"
+#include "envoy/network/address.h"
+
+#include "common/network/cidr_range.h"
+#include "common/network/utility.h"
+
+#include "absl/numeric/int128.h"
+#include "fmt/format.h"
+
+namespace Envoy {
+namespace Network {
+namespace LcTrie {
+
+/**
+ * Level Compressed Trie implemenation for IP(Ipv4 or IPv6) Prefixes.
+ */
+class LcTrie {
+public:
+  /**
+   * @param fill_factor supplies the fraction of completeness to use when considering
+   *        if a subtrie should be compressed.
+   * @param root_branching_factor supplies the branching factor at the root. This is
+   *        an optimization as suggested in the paper.
+   * @param tag_data supplies the ip tag and the prefixes for associated with that tag.
+   */
+  LcTrie(const std::vector<std::pair<std::string, std::vector<Address::CidrRange>>>& tag_data,
+         double fill_factor = 0.5, uint32_t root_branching_factor = 0);
+
+  /**
+   * Search the LC Trie for a prefix that has the input, ip, in range.
+   * @param  ip supplies the IP address searching for a prefix in string format.
+   * @return tag from the prefix that encompasses the input. An empty string is returned
+   *         if no prefix in the LC Trie exists.
+   */
+  std::string search(Network::Address::InstanceConstSharedPtr ip_address) const;
+
+private:
+  /**
+   * Extract n bits from input starting at position p.
+   * @param position supplies the position.
+   * @param n supplies the number of bits to extract.
+   * @param input supplies the
+   */
+  template <class IpType, uint32_t address_size = 8 * sizeof(IpType)>
+  static IpType extractBits(int p, int n, IpType input) {
+    // The IP's are stored in network byte order. By shifting the value to the left by p bits, the
+    // bits between 0 and p-1 are zero'd out. Then to get the n bits, shift the IP back by the
+    // address_size minus the number of desired bits.
+    return input << p >> (address_size - n);
```

Comment:
I think this may be unsafe. C doesn't specify whether a right shift is logical (high order bits get filled with zeros) or logical (high order bits get filled with a copy of whatever the input high order bit was).

It's probably safest to bitwise-and the result of these shifts with a mask consisting of _n_ ones in the low order bits.

Reply:
I double checked the c++ standard and for unsigned left/right shifting is safe and produces the values we expect. In this version of the pr, they were int's. I've change the methods to be uint32_t.